File: backend/app/schemas/arcade.py
from pydantic import BaseModel


class QuizItem(BaseModel):
    lesson_id: str
    question: str
    choices: list[str]
    answer_index: int


class QuizSessionOut(BaseModel):
    items: list[QuizItem]


class QuizAnswer(BaseModel):
    lesson_id: str
    choice_index: int
    time_ms: int = 0


class QuizScoreIn(BaseModel):
    session_items: list[QuizItem]
    answers: list[QuizAnswer]


class QuizScoreOut(BaseModel):
    points: int
    coins_awarded: int
    personal_best: int
    leaderboard_rank: int | None


class LeaderboardEntryOut(BaseModel):
    username: str
    country_code: str
    points: int


class LeaderboardOut(BaseModel):
    entries: list[LeaderboardEntryOut]

File: frontend/src/api/arcade.ts
import { useQuery } from '@tanstack/react-query';
import { apiFetch } from './client';

export type QuizItem = { lesson_id: string; question: string; choices: string[]; answer_index: number };
export type QuizSession = { items: QuizItem[] };
export type QuizAnswer = { lesson_id: string; choice_index: number; time_ms: number };
export type QuizScoreBody = { session_items: QuizItem[]; answers: QuizAnswer[] };
export type QuizScoreResult = {
  points: number; coins_awarded: number; personal_best: number; leaderboard_rank: number | null;
};
export type ArcadeLeaderboard = { entries: { username: string; country_code: string; points: number }[] };

export function getQuizRushSession() {
  return apiFetch<QuizSession>('/arcade/quiz-rush/session');
}

export function submitQuizRushScore(body: QuizScoreBody) {
  return apiFetch<QuizScoreResult>('/arcade/quiz-rush/score', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function getArcadeLeaderboard(game = 'quiz_rush') {
  return apiFetch<ArcadeLeaderboard>(`/arcade/leaderboard?game=${encodeURIComponent(game)}`);
}

export function useArcadeLeaderboard(game = 'quiz_rush') {
  return useQuery({ queryKey: ['arcade', 'leaderboard', game], queryFn: () => getArcadeLeaderboard(game) });
}
